File: benchmarks/matlab/scc2_memory.m
function scc2_memory(input_dir, requested_threads, stage, rhs_count, output_file, source_commit, source_archive_sha256)
%SCC2_MEMORY Stop at an exact process-memory checkpoint for SCC2.
arguments
    input_dir (1,:) char
    requested_threads (1,1) double {mustBeInteger,mustBePositive}
    stage (1,:) char {mustBeMember(stage, {'baseline','input','graph','hierarchy','plan','workspace-one','workspace-pool','solve','batch'})}
    rhs_count (1,1) double {mustBeInteger,mustBePositive}
    output_file (1,:) char
    source_commit (1,:) char
    source_archive_sha256 (1,:) char
end

maxNumCompThreads(requested_threads);
started = tic;
checkpoints = checkpoint('baseline', started, 0);
family = getenv_default('CMG_FAMILY', 'unknown');
vertices = str2double(getenv_default('CMG_VERTICES', '0'));
edge_count = 0;
matrix_nonzeros = 0;
owned = struct('input', 0, 'graph', 0, 'hierarchy', 0, 'plan', 0, 'workspace_one', 0, 'workspace_pool', 0);
numerical = [];
warnings = {};
if strcmp(stage, 'baseline')
    write_result();
    return
end

[vertices, endpoints_u, endpoints_v, weights] = read_graph(fullfile(input_dir, 'graph.bin'));
right_hand_sides = read_vectors(fullfile(input_dir, 'rhs.bin'));
truths = read_vectors(fullfile(input_dir, 'truth.bin')); %#ok<NASGU>
metadata = jsondecode(fileread(fullfile(input_dir, 'metadata.json')));
family = metadata.family;
edge_count = numel(weights);
visible = whos('endpoints_u', 'endpoints_v', 'weights', 'right_hand_sides', 'truths');
owned.input = sum([visible.bytes]);
checkpoints(end + 1) = checkpoint('input', started, owned.input); %#ok<AGROW>
assert(rhs_count <= size(right_hand_sides, 2), 'Fixture has too few RHSs');
if strcmp(stage, 'input')
    write_result();
    return
end

u = double(endpoints_u) + 1;
v = double(endpoints_v) + 1;
diagonal = accumarray([u; v], [weights; weights], [vertices 1], @sum, 0);
A = sparse([u; v; (1:vertices)'], [v; u; (1:vertices)'], [-weights; -weights; diagonal], vertices, vertices);
matrix_nonzeros = nnz(A);
visible = whos('A');
owned.graph = sum([visible.bytes]);
checkpoints(end + 1) = checkpoint('graph', started, owned.graph); %#ok<AGROW>
if strcmp(stage, 'graph')
    write_result();
    return
end

opts = struct('validate', 0, 'matlab_', 0);
[pfun, H, hierarchy_flag] = cmg_precondition(A, opts);
visible = whos('H');
owned.hierarchy = sum([visible.bytes]);
checkpoints(end + 1) = checkpoint('hierarchy', started, owned.hierarchy); %#ok<AGROW>
if hierarchy_flag ~= 0
    warnings = {sprintf('Official CMG hierarchy flag %g', hierarchy_flag)};
end
if strcmp(stage, 'hierarchy')
    write_result();
    return
end

% The official workflow has no separately retained parallel plan.
warnings{end + 1} = 'parallel-plan stage unsupported by official MATLAB workflow';
checkpoints(end + 1) = checkpoint('plan', started, 0); %#ok<AGROW>
if strcmp(stage, 'plan')
    write_result();
    return
end

workspace_one = zeros(vertices, 1);
visible = whos('workspace_one');
owned.workspace_one = sum([visible.bytes]);
checkpoints(end + 1) = checkpoint('workspace-one', started, owned.workspace_one); %#ok<AGROW>
if strcmp(stage, 'workspace-one')
    write_result();
    return
end

workspace_pool = zeros(vertices, rhs_count); %#ok<NASGU>
visible = whos('workspace_pool');
owned.workspace_pool = sum([visible.bytes]);
checkpoints(end + 1) = checkpoint('workspace-pool', started, owned.workspace_pool); %#ok<AGROW>
if strcmp(stage, 'workspace-pool')
    write_result();
    return
end

if strcmp(stage, 'solve')
    [solution, flag, ~, iterations] = pcg(A, right_hand_sides(:, 1), 1e-8, 1000, pfun, [], workspace_one);
    residual = right_hand_sides(:, 1) - A * solution;
    operator_bound = 2 * max(full(diag(A)));
    backward = norm(residual) / max(norm(right_hand_sides(:, 1)) + operator_bound * norm(solution), realmin);
    numerical = struct('iterations', iterations, 'max_backward_error', backward, 'native_flag', flag);
    assert(flag == 0 && backward <= 1.1e-8, 'Memory-stage solve failed');
    checkpoints(end + 1) = checkpoint('solve', started, 0); %#ok<AGROW>
else
    max_backward = 0;
    total_iterations = 0;
    for rhs_index = 1:rhs_count
        [solution, flag, ~, iterations] = pcg(A, right_hand_sides(:, rhs_index), 1e-8, 1000, pfun, [], workspace_one);
        residual = right_hand_sides(:, rhs_index) - A * solution;
        operator_bound = 2 * max(full(diag(A)));
        backward = norm(residual) / max(norm(right_hand_sides(:, rhs_index)) + operator_bound * norm(solution), realmin);
        assert(flag == 0 && backward <= 1.1e-8, 'Memory-stage batch failed');
        max_backward = max(max_backward, backward);
        total_iterations = total_iterations + iterations;
    end
    numerical = struct('iterations', total_iterations, 'max_backward_error', max_backward, 'native_flag', 0);
    checkpoints(end + 1) = checkpoint('batch', started, 0); %#ok<AGROW>
end
write_result();

    function write_result()
        result = struct();
        result.protocol_version = 'cmg-scc2-v1';
        result.record_type = 'memory-stage';
        result.run_id = getenv_default('CMG_RUN_ID', 'local');
        result.task_id = str2double(getenv_default('CMG_TASK_ID', '1'));
        result.source_commit = source_commit;
        result.source_archive_sha256 = source_archive_sha256;
        result.binary_sha256 = getenv_default('CMG_MEX_BINARY_SHA256', repmat('0', 1, 64));
        result.environment_id = getenv_default('CMG_ENVIRONMENT_ID', repmat('0', 1, 64));
        result.implementation = 'matlab';
        result.family = family;
        result.vertices = vertices;
        result.canonical_edges = edge_count;
        result.matrix_nonzeros = matrix_nonzeros;
        result.threads = maxNumCompThreads;
        result.rhs_count = rhs_count;
        result.stage = stage;
        result.owned_bytes = owned;
        result.checkpoints = checkpoints;
        result.numerical = numerical;
        result.peak_rss_kb = peak_rss_kb();
        result.warnings = warnings;
        result.success = true;
        parent = fileparts(output_file);
        if ~isempty(parent) && ~isfolder(parent)
            mkdir(parent);
        end
        temporary = [output_file '.tmp'];
        fid = fopen(temporary, 'w');
        assert(fid >= 0, 'Could not open output file');
        cleanup = onCleanup(@() fclose(fid));
        fprintf(fid, '%s\n', jsonencode(result, PrettyPrint=true));
        clear cleanup
        movefile(temporary, output_file, 'f');
        fprintf('CMG_SCC2_MATLAB_MEMORY_SUCCESS stage=%s threads=%d\n', stage, result.threads);
    end
end

function value = checkpoint(stage, started, owned_bytes)
value = struct('stage', stage, 'elapsed_ns', round(toc(started) * 1e9), 'peak_rss_kb', peak_rss_kb(), 'owned_bytes', owned_bytes);
end

function value = getenv_default(name, fallback)
value = getenv(name);
if isempty(value)
    value = fallback;
end
end

function value = peak_rss_kb()
value = 0;
try
    pid = feature('getpid');
    text = fileread(sprintf('/proc/%d/status', pid));
    token = regexp(text, 'VmHWM:\s+(\d+)\s+kB', 'tokens', 'once');
    if ~isempty(token)
        value = str2double(token{1});
    end
catch
    value = 0;
end
end

function [vertices, u, v, weights] = read_graph(path)
fid = fopen(path, 'r', 'ieee-le');
assert(fid >= 0, 'Could not open graph input');
cleanup = onCleanup(@() fclose(fid));
magic = char(fread(fid, 8, 'uint8=>uint8')');
assert(strcmp(magic, 'CMGGRPH1'), 'Invalid graph magic');
vertices = double(fread(fid, 1, 'uint64=>uint64'));
edge_count = double(fread(fid, 1, 'uint64=>uint64'));
records = fread(fid, [3 edge_count], 'uint64=>uint64');
assert(size(records, 2) == edge_count, 'Truncated graph input');
u = records(1, :)';
v = records(2, :)';
weights = reshape(typecast(records(3, :), 'double'), [], 1);
end

function values = read_vectors(path)
fid = fopen(path, 'r', 'ieee-le');
assert(fid >= 0, 'Could not open vector input');
cleanup = onCleanup(@() fclose(fid));
magic = char(fread(fid, 8, 'uint8=>uint8')');
assert(strcmp(magic, 'CMGVEC01'), 'Invalid vector magic');
dimension = double(fread(fid, 1, 'uint64=>uint64'));
count = double(fread(fid, 1, 'uint64=>uint64'));
values = fread(fid, [dimension count], 'double=>double');
assert(isequal(size(values), [dimension count]), 'Truncated vector input');
end

File: benchmarks/src/bin/scc2-memory.rs
use cmg::{
    CmgOptions, CmgPreconditioner, Laplacian, ParallelCmgPlan, ParallelExecutor, ParallelOptions,
    PcgOptions, PcgWorkspace, solve_pcg_batch_with_executor, solve_pcg_with_workspace,
};
use std::env;
use std::error::Error;
use std::fs::{self, File};
use std::io::{self, BufReader, Read, Write};
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::Instant;

type AnyError = Box<dyn Error>;
type EdgeSpec = (usize, usize, f64);

const GRAPH_MAGIC: &[u8; 8] = b"CMGGRPH1";
const VECTOR_MAGIC: &[u8; 8] = b"CMGVEC01";
const SOURCE_COMMIT: &str = match option_env!("CMG_BENCH_COMMIT") {
    Some(value) => value,
    None => "unknown",
};
const SOURCE_ARCHIVE_SHA256: &str = match option_env!("CMG_BENCH_ARCHIVE_SHA256") {
    Some(value) => value,
    None => "unknown",
};

fn main() -> Result<(), AnyError> {
    let mut arguments = env::args().skip(1);
    let input = PathBuf::from(required(&mut arguments, "input-dir")?);
    let threads = required(&mut arguments, "threads")?.parse::<usize>()?;
    let stage = required(&mut arguments, "stage")?;
    let rhs_count = required(&mut arguments, "rhs-count")?.parse::<usize>()?;
    let output = PathBuf::from(required(&mut arguments, "output")?);
    if arguments.next().is_some() || threads == 0 || rhs_count == 0 {
        return Err(io::Error::new(io::ErrorKind::InvalidInput, "invalid arguments").into());
    }
    if !matches!(
        stage.as_str(),
        "baseline"
            | "input"
            | "graph"
            | "hierarchy"
            | "plan"
            | "workspace-one"
            | "workspace-pool"
            | "solve"
            | "batch"
    ) {
        return Err(io::Error::new(io::ErrorKind::InvalidInput, "invalid stage").into());
    }
    run(&input, threads, &stage, rhs_count, &output)
}

fn run(
    input: &Path,
    threads: usize,
    stop: &str,
    rhs_count: usize,
    output: &Path,
) -> Result<(), AnyError> {
    let started = Instant::now();
    let mut checkpoints = vec![checkpoint("baseline", started, 0)];
    let family = env::var("CMG_FAMILY").unwrap_or_else(|_| "unknown".to_owned());
    let vertices_from_env = env::var("CMG_VERTICES")
        .ok()
        .and_then(|value| value.parse::<usize>().ok())
        .unwrap_or(0);
    if stop == "baseline" {
        return finish(
            output,
            stop,
            threads,
            rhs_count,
            &family,
            vertices_from_env,
            0,
            0,
            0,
            0,
            0,
            0,
            checkpoints,
            None,
        );
    }

    let (vertices, edges) = read_graph(&input.join("graph.bin"))?;
    let right_hand_sides = read_vectors(&input.join("rhs.bin"))?;
    let _truths = read_vectors(&input.join("truth.bin"))?;
    if rhs_count > right_hand_sides.len() {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "too few RHSs").into());
    }
    let input_bytes = edges.len() * std::mem::size_of::<EdgeSpec>()
        + right_hand_sides
            .iter()
            .map(|value| std::mem::size_of_val(value.as_slice()))
            .sum::<usize>();
    checkpoints.push(checkpoint("input", started, input_bytes));
    if stop == "input" {
        return finish(
            output,
            stop,
            threads,
            rhs_count,
            &family,
            vertices,
            edges.len(),
            0,
            input_bytes,
            0,
            0,
            0,
            checkpoints,
            None,
        );
    }

    let graph = Laplacian::from_edges(vertices, edges.iter().copied())?;
    let graph_bytes = graph.edge_count() * std::mem::size_of::<cmg::Edge>()
        + std::mem::size_of_val(graph.diagonal());
    checkpoints.push(checkpoint("graph", started, graph_bytes));
    if stop == "graph" {
        return finish(
            output,
            stop,
            threads,
            rhs_count,
            &family,
            vertices,
            graph.edge_count(),
            graph.matrix_nnz(),
            input_bytes,
            graph_bytes,
            0,
            0,
            checkpoints,
            None,
        );
    }

    let executor = ParallelExecutor::new(ParallelOptions {
        threads,
        min_parallel_len: 16_384,
        ..ParallelOptions::default()
    })?;
    let preconditioner =
        CmgPreconditioner::build_with_executor(&graph, CmgOptions::default(), &executor)?;
    let hierarchy_bytes = hierarchy_bytes(&preconditioner);
    checkpoints.push(checkpoint("hierarchy", started, hierarchy_bytes));
    if stop == "hierarchy" {
        return finish(
            output,
            stop,
            threads,
            rhs_count,
            &family,
            vertices,
            graph.edge_count(),
            graph.matrix_nnz(),
            input_bytes,
            graph_bytes,
            hierarchy_bytes,
            0,
            checkpoints,
            None,
        );
    }

    let plan = ParallelCmgPlan::build(&preconditioner, &executor)?;
    let plan_bytes = plan.byte_len();
    checkpoints.push(checkpoint("plan", started, plan_bytes));
    if stop == "plan" {
        return finish(
            output,
            stop,
            threads,
            rhs_count,
            &family,
            vertices,
            graph.edge_count(),
            graph.matrix_nnz(),
            input_bytes,
            graph_bytes,
            hierarchy_bytes,
            plan_bytes,
            checkpoints,
            None,
        );
    }

    let one_workspace = PcgWorkspace::new(&preconditioner);
    let workspace_bytes = one_workspace.byte_len();
    checkpoints.push(checkpoint("workspace-one", started, workspace_bytes));
    if stop == "workspace-one" {
        return finish(
            output,
            stop,
            threads,
            rhs_count,
            &family,
            vertices,
            graph.edge_count(),
            graph.matrix_nnz(),
            input_bytes,
            graph_bytes,
            hierarchy_bytes,
            plan_bytes,
            checkpoints,
            None,
        );
    }

    let concurrency = executor.batch_concurrency(workspace_bytes, rhs_count)?;
    let mut workspace_pool = (0..concurrency)
        .map(|_| PcgWorkspace::new(&preconditioner))
        .collect::<Vec<_>>();
    checkpoints.push(checkpoint(
        "workspace-pool",
        started,
        workspace_bytes.saturating_mul(concurrency),
    ));
    if stop == "workspace-pool" {
        return finish(
            output,
            stop,
            threads,
            rhs_count,
            &family,
            vertices,
            graph.edge_count(),
            graph.matrix_nnz(),
            input_bytes,
            graph_bytes,
            hierarchy_bytes,
            plan_bytes,
            checkpoints,
            None,
        );
    }

    let options = PcgOptions {
        relative_tolerance: 1.0e-8,
        max_iterations: 1_000,
        ..PcgOptions::default()
    };
    let numerical = if stop == "solve" {
        let result = solve_pcg_with_workspace(
            &graph,
            &preconditioner,
            &right_hand_sides[0],
            options,
            &mut workspace_pool[0],
        )?;
        checkpoints.push(checkpoint("solve", started, 0));
        Some((result.iterations(), result.backward_error()))
    } else {
        let results = solve_pcg_batch_with_executor(
            &graph,
            &preconditioner,
            &right_hand_sides[..rhs_count],
            options,
            &executor,
        )?;
        checkpoints.push(checkpoint("batch", started, 0));
        Some((
            results.iter().map(|value| value.iterations()).sum(),
            results
                .iter()
                .map(|value| value.backward_error())
                .fold(0.0, f64::max),
        ))
    };
    finish(
        output,
        stop,
        threads,
        rhs_count,
        &family,
        vertices,
        graph.edge_count(),
        graph.matrix_nnz(),
        input_bytes,
        graph_bytes,
        hierarchy_bytes,
        plan_bytes,
        checkpoints,
        numerical,
    )
}

#[allow(clippy::too_many_arguments)]
fn finish(
    output: &Path,
    stage: &str,
    threads: usize,
    rhs_count: usize,
    family: &str,
    vertices: usize,
    edges: usize,
    matrix_nonzeros: usize,
    input_bytes: usize,
    graph_bytes: usize,
    hierarchy_bytes: usize,
    plan_bytes: usize,
    checkpoints: Vec<String>,
    numerical: Option<(usize, f64)>,
) -> Result<(), AnyError> {
    let binary_sha256 = sha256sum(&env::current_exe()?)?;
    let numerical = numerical.map_or_else(
        || "null".to_owned(),
        |(iterations, backward)| {
            format!("{{\"iterations\":{iterations},\"max_backward_error\":{backward:.17e}}}")
        },
    );
    let payload = format!(
        concat!(
            "{{\"protocol_version\":\"cmg-scc2-v1\",",
            "\"record_type\":\"memory-stage\",\"run_id\":{},\"task_id\":{},",
            "\"source_commit\":{},\"source_archive_sha256\":{},\"binary_sha256\":{},",
            "\"environment_id\":{},\"implementation\":\"rust\",\"family\":{},",
            "\"vertices\":{},\"canonical_edges\":{},\"matrix_nonzeros\":{},",
            "\"threads\":{},\"rhs_count\":{},\"stage\":{},",
            "\"owned_bytes\":{{\"input\":{},\"graph\":{},\"hierarchy\":{},\"plan\":{}}},",
            "\"checkpoints\":[{}],\"numerical\":{},\"peak_rss_kb\":{},\"success\":true}}\n"
        ),
        json_string(&env::var("CMG_RUN_ID").unwrap_or_else(|_| "local".to_owned())),
        env::var("CMG_TASK_ID")
            .ok()
            .and_then(|value| value.parse::<usize>().ok())
            .unwrap_or(1),
        json_string(SOURCE_COMMIT),
        json_string(SOURCE_ARCHIVE_SHA256),
        json_string(&binary_sha256),
        json_string(&env::var("CMG_ENVIRONMENT_ID").unwrap_or_else(|_| "unknown".to_owned())),
        json_string(family),
        vertices,
        edges,
        matrix_nonzeros,
        threads,
        rhs_count,
        json_string(stage),
        input_bytes,
        graph_bytes,
        hierarchy_bytes,
        plan_bytes,
        checkpoints.join(","),
        numerical,
        peak_rss_kb(),
    );
    atomic_write(output, payload.as_bytes())?;
    println!("CMG_SCC2_MEMORY_SUCCESS stage={stage} threads={threads}");
    Ok(())
}

fn checkpoint(stage: &str, started: Instant, owned_bytes: usize) -> String {
    format!(
        "{{\"stage\":{},\"elapsed_ns\":{},\"peak_rss_kb\":{},\"owned_bytes\":{owned_bytes}}}",
        json_string(stage),
        started.elapsed().as_nanos(),
        peak_rss_kb(),
    )
}

fn hierarchy_bytes(preconditioner: &CmgPreconditioner) -> usize {
    preconditioner
        .hierarchy()
        .levels()
        .iter()
        .map(|level| {
            level.graph().edge_count() * std::mem::size_of::<cmg::Edge>()
                + std::mem::size_of_val(level.graph().diagonal())
                + std::mem::size_of_val(level.inverse_diagonal())
                + level.aggregation().map_or(0, |aggregation| {
                    std::mem::size_of_val(aggregation.labels())
                        + std::mem::size_of_val(aggregation.sizes())
                })
        })
        .sum::<usize>()
        + preconditioner.component_metadata_bytes()
        + preconditioner
            .terminal_factor()
            .map_or(0, |factor| factor.byte_len())
}

fn peak_rss_kb() -> u64 {
    fs::read_to_string("/proc/self/status")
        .ok()
        .and_then(|content| {
            content.lines().find_map(|line| {
                line.strip_prefix("VmHWM:")
                    .and_then(|value| value.split_whitespace().next())
                    .and_then(|value| value.parse().ok())
            })
        })
        .unwrap_or(0)
}

fn read_graph(path: &Path) -> Result<(usize, Vec<EdgeSpec>), AnyError> {
    let mut reader = BufReader::new(File::open(path)?);
    let mut magic = [0_u8; 8];
    reader.read_exact(&mut magic)?;
    if &magic != GRAPH_MAGIC {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid graph magic").into());
    }
    let vertices = usize::try_from(read_u64(&mut reader)?)?;
    let count = usize::try_from(read_u64(&mut reader)?)?;
    let mut edges = Vec::with_capacity(count);
    for _ in 0..count {
        edges.push((
            usize::try_from(read_u64(&mut reader)?)?,
            usize::try_from(read_u64(&mut reader)?)?,
            f64::from_bits(read_u64(&mut reader)?),
        ));
    }
    Ok((vertices, edges))
}

fn read_vectors(path: &Path) -> Result<Vec<Vec<f64>>, AnyError> {
    let mut reader = BufReader::new(File::open(path)?);
    let mut magic = [0_u8; 8];
    reader.read_exact(&mut magic)?;
    if &magic != VECTOR_MAGIC {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid vector magic").into());
    }
    let dimension = usize::try_from(read_u64(&mut reader)?)?;
    let count = usize::try_from(read_u64(&mut reader)?)?;
    let mut values = vec![vec![0.0; dimension]; count];
    for vector in &mut values {
        for value in vector {
            *value = f64::from_bits(read_u64(&mut reader)?);
        }
    }
    Ok(values)
}

fn read_u64(reader: &mut impl Read) -> Result<u64, io::Error> {
    let mut bytes = [0_u8; 8];
    reader.read_exact(&mut bytes)?;
    Ok(u64::from_le_bytes(bytes))
}

fn required(arguments: &mut impl Iterator<Item = String>, name: &str) -> Result<String, AnyError> {
    arguments.next().ok_or_else(|| {
        io::Error::new(io::ErrorKind::InvalidInput, format!("missing {name}")).into()
    })
}

fn sha256sum(path: &Path) -> Result<String, AnyError> {
    let output = Command::new("sha256sum").arg(path).output()?;
    if !output.status.success() {
        return Err(io::Error::other("sha256sum failed").into());
    }
    String::from_utf8(output.stdout)?
        .split_whitespace()
        .next()
        .map(str::to_owned)
        .ok_or_else(|| io::Error::other("sha256sum produced no digest").into())
}

fn json_string(value: &str) -> String {
    format!("{:?}", value)
}

fn atomic_write(path: &Path, content: &[u8]) -> Result<(), io::Error> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    let temporary = path.with_extension("tmp");
    let mut file = File::create(&temporary)?;
    file.write_all(content)?;
    file.sync_all()?;
    fs::rename(temporary, path)
}
